src/Guard.php

```diff
@@ -215,131 +205,126 @@ public function generateToken()
-        if (!$this->persistentTokenMode) {
-            $this->removeFromStorage($name);
-        }
+        return $this->keyPair[$this->getTokenNameKey()] ?? null;
+    }
 
-        return $result;
+    /**
+     * @return string
+     */
+    public function getTokenValue(): ?string
+    {
+        return $this->keyPair[$this->getTokenValueKey()] ?? null;
     }
 
     /**
-     * Create CSRF token value
-     *
      * @return string
      */
-    protected function createToken()
+    public function getTokenNameKey(): string
     {
-        return bin2hex(random_bytes($this->strength));
+        return $this->prefix . '_name';
     }
 
     /**
-     * Save token to storage
-     *
-     * @param  string $name  CSRF token name
-     * @param  string $value CSRF token value
+     * @return string
      */
-    protected function saveToStorage($name, $value)
+    public function getTokenValueKey(): string
     {
-        $this->storage[$name] = $value;
+        return $this->prefix . '_value';
     }
 
     /**
-     * Get token from storage
-     *
-     * @param  string      $name CSRF token name
-     *
-     * @return string|bool CSRF token value or `false` if not present
+     * @return bool
      */
-    protected function getFromStorage($name)
+    public function getPersistentTokenMode(): bool
     {
-        return isset($this->storage[$name]) ? $this->storage[$name] : false;
+        return $this->persistentTokenMode;
     }
 
     /**
-     * Get the most recent key pair from storage.
-     *
-     * @return string[]|null Array containing name and value if found, null otherwise
-     */    
-    protected function getLastKeyPair()
+     * @return string[]|null
+     */
+    protected function getLastKeyPair(): ?array
     {
         // Use count, since empty ArrayAccess objects can still return false for `empty`
         if (count($this->storage) < 1) {
             return null;
         }
 
-        foreach ($this->storage as $name => $value) {
-            continue;
+        $name = null;
+        $value = null;
+        foreach ($this->storage as $k => $v) {
+            $name = $k;
+            $value = $v;
         }
 
-        $keyPair = [
-            $this->prefix . '_name' => $name,
-            $this->prefix . '_value' => $value
-        ];
-
-        return $keyPair;
+        return $name !== null && $value !== null
+            ? [
+                $this->prefix . '_name' => $name,
+                $this->prefix . '_value' => $value
+            ]
+            : null;
     }
-    
+
     /**
      * Load the most recent key pair in storage.
      *
      * @return bool `true` if there was a key pair to load in storage, false otherwise.
      */
-    protected function loadLastKeyPair()
+    protected function loadLastKeyPair(): bool
     {
-        $this->keyPair = $this->getLastKeyPair();
-
-        if ($this->keyPair) {
-            return true;
-        }
+        return !!($this->keyPair = $this->getLastKeyPair());
```

Comment:
Are double exclamation marks better than a simple typecast?
```php
        return (bool)($this->keyPair = $this->getLastKeyPair());
```

Reply:
will change to typecast since it's more readable